File: SOLID-Java/src/main/test/dependencyinversion/before/DatabaseHandlerTest.java
package dependencyinversion.before;

import org.junit.Before;
import org.junit.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.MockitoAnnotations;

public class DatabaseHandlerTest {

	@Mock
	MySQLDatabase mySQLDb;

	@Mock
	OracleDatabase oracleDb;

	@InjectMocks
	private DatabaseHandler handler = new DatabaseHandler();

	@Before
	public void setup() {
		MockitoAnnotations.initMocks(this);
	}

	@Test
	public void testMySQL() throws Exception {
		handler.connect(mySQLDb);

		Mockito.verify(mySQLDb, Mockito.times(1)).connect();
	}

	@Test
	public void testMyOracle() throws Exception {
		handler.connect(oracleDb);
		Mockito.verify(oracleDb, Mockito.times(1)).connect();
	}

}

File: SOLID-Scala/src/main/test/dependencyinversion/after/DatabaseHandlerTest.scala
package dependencyinversion.after

import org.scalatest.BeforeAndAfter
import org.scalatest.FunSuite

class DatabaseHandlerTest extends FunSuite with BeforeAndAfter {

  private val handler = new DatabaseHandler()

  test("MySQL DB") {
    val dbConnect = handler.connect(new MySQLDatabase)
    val dbDisconnect = handler.disconnect(new MySQLDatabase)
    assert(dbConnect.equals("MySQL"))
    assert(dbDisconnect.equals("MySQL"))
  }
  
   test("Oracle DB") {
    val dbConnect = handler.connect(new OracleDatabase)
    val dbDisconnect = handler.disconnect(new OracleDatabase)
    assert(dbConnect.equals("Oracle"))
    assert(dbDisconnect.equals("Oracle"))
  }

}